File: rubymacros/make_cpw.rb
# $description: Make CPW out of Path
# $autorun
# $show-in-menu
class MenuAction < RBA::Action
  def initialize( title, shortcut, &action ) 
    self.title = title
    self.shortcut = shortcut
    @action = action
  end
  def triggered 
    @action.call( self ) 
  end
private
  @action
end

$make_cpw = MenuAction.new( "Make CPW", "" ) do 
  app = Application.instance
  mw = app.main_window
  lv = mw.current_view
  raise "No view selected" if lv.nil?
  ly = lv.active_cellview.layout
  dbu = ly.dbu
  
  ri,ro,n = 95/dbu,105/dbu,10/dbu # Divide by dbu to convert from um to database units (which are integers)
  
  cpw_layer = ly.layer(1,0)
  keepout_layer= ly.layer(1,6)
  
  lv.transaction("Convert Paths to CPW")

  begin
  
    lv.each_object_selected { |obj|
  
      if !obj.is_cell_inst?
      # Define new paths
       inner = Path.new()
       outer = Path.new()
       keepout = Path.new()
       inner.assign(obj.shape.path)
       outer.assign(obj.shape.path)
       keepout.assign(obj.shape.path)

      # Make outer path wider
       outer.width = 22.0/dbu
       
       #Make keepout wider
       keepout.width = 30.0/dbu
      
      #Convert to polygon and round corners
       inner = [inner.polygon.round_corners(ri,ro,n)]
       outer = [outer.polygon.round_corners(ri-(6/dbu), ro+(6/dbu), n)]
       keepout = keepout.polygon.round_corners(ri-(10/dbu), ro+(10/dbu), n)
      
       ep = RBA::EdgeProcessor::new
       outer_minus_inner = ep.boolean_p2p(outer, inner, RBA::EdgeProcessor::ModeANotB, true, false)

      #Insert cpw into correct layer
      outer_minus_inner.each do |p|
       lv.active_cellview.cell.shapes(cpw_layer).insert(p)
       end

      #Insert keepout into correct layer
      lv.active_cellview.cell.shapes(keepout_layer).insert(keepout)
       
      end  
    }
  
  ensure
    lv.commit
  end
end

app = RBA::Application.instance
mw = app.main_window

menu = mw.menu
menu.insert_item("tools_menu.end", "make_cpw", $make_cpw)

File: pymacros/Make CPW.py
# $description: Make CPW out of Path
# $autorun
# $show-in-menu
import pya
import sys
sys.stderr = sys.stdout

class MenuAction(pya.Action):
    def __init__(self, title, shortcut, action):
        self.title = title
        self.shortcut = shortcut
        self.action = action
        
    def triggered(self):
        self.action()

def make_cpw():

    #Load View
    app = pya.Application.instance()
    mw = app.main_window()
    lv = mw.current_view()
    ly = lv.active_cellview().layout()
    dbu = ly.dbu
    
    center_radius = 100.0/dbu #Curve radius
    center_width = 10./dbu #Center conductor width
    s_width = 6./dbu #Center - Outer conductor spacing
    keepout_width = 4./dbu #Keepout spacing
    
    ri = center_radius - .5*center_width #Inner radius
    ro = center_radius + .5*center_width # Outer radius
    n = 10/dbu #Number of points per curve
    
    cpw_layer = ly.layer(1,0)
    keepout_layer = ly.layer(1,6)
    
    if lv==None:
        raise Exception("No view selected")
    
    lv.transaction("Make CPW")
    
    selected_objects = lv.each_object_selected()
    print("x")
    for obj in selected_objects:
        if obj.shape.is_path() and not obj.is_cell_inst():
            inner = pya.Path()
            outer = pya.Path()
            keepout = pya.Path()
            print(obj.shape.path)
            
            inner = obj.shape.path
            outer = obj.shape.path
            keepout = obj.shape.path
            
            #Adjust widths
            inner.width = center_width
            outer.width = inner.width + 2*s_width
            keepout.width = outer.width + 2*keepout_width
            
            #Round Corners
            inner = [inner.polygon().round_corners(ri, ro, n)]
            outer = [outer.polygon().round_corners(ri-s_width,ro+s_width, n)]
            keepout = keepout.polygon().round_corners(ri-s_width-keepout_width, ro+s_width+keepout_width, n)
            
            #Subtract inner from outer
            ep = pya.EdgeProcessor()
            outer_minus_inner = ep.boolean_p2p(outer, inner, pya.EdgeProcessor().ModeANotB, True, False)
            
            for p in outer_minus_inner:
                lv.active_cellview().cell.shapes(cpw_layer).insert(p)
            
           # lv.active_cellview().cell.shapes(cpw_layer).insert(outer_minus_inner[0])
            lv.active_cellview().cell.shapes(keepout_layer).insert(keepout)
            
    lv.commit()
    
    #pya.MessageBox.info("Test", "It worked bitches", pya.MessageBox.b_ok())
    
x = MenuAction("Make CPW", "", make_cpw)

app = pya.Application.instance()
mw = app.main_window()
menu = mw.menu()
menu.insert_separator("tools_menu.end", "sep_cpw")
menu.insert_item("tools_menu.end", "cpw", x)